02-peer-protocol.md

```diff
@@ -894,40 +916,43 @@ any message), they are independent of requirements here.
 ### Requirements
 
 A node MUST handle continuing a previous channel on a new encrypted
-transport.  On disconnection, a node MAY forget nodes which have not
-sent or received an `accept_channel` message, and MAY forget nodes
-which have not sent `funding_locked` after a reasonable timeout.
+transport.
+
+On disconnection, the funder MUST remember the channel for
+reconnection if it has broadcast the funding transaction, otherwise it
+MUST NOT.
+
+On disconnection, the non-funding node MUST remember the channel for
+reconnection if it has sent the `funding_signed` message, otherwise
+it MUST NOT.
 
 On disconnection, a node MUST reverse any uncommitted updates sent by
 the other side (ie. all messages beginning with `update_` for which no
 `commitment_signed` has been received).  Note that a node MAY have
 already use the `payment_preimage` value from the `update_fulfill_htlc`,
 so the effects of `update_fulfill_htlc` is not completely reversed.
 
-On reconnection, a node MUST retransmit old messages which may not
+On reconnection, if a channel is in an error state, the node SHOULD
+retransmit the error packet and ignore any other packets for that
+channel, or if the channel has entered closing negotiation, the node
+MUST retransmit the last `closing_signed`.
+
+Otherwise, a node MUST retransmit old messages after `funding_signed` which may not
 have been received, and MUST NOT retransmit old messages which have
 been explicitly or implicitly acknowledged.  The following table
 lists the acknowledgment conditions for each message:
 
-* `open_channel`: acknowledged by `accept_channel`.
-* `accept_channel`: acknowledged by `funding_created`.
-* `funding_created`: acknowledged by `funding_signed`.
-* `funding_signed`: acknowledged by `funding_locked`.
 * `funding_locked`: acknowledged by `update_` messages, `commitment_signed`, `revoke_and_ack` or `shutdown` messages.
 * `update_` messages: acknowledged by `revoke_and_ack`.
 * `commitment_signed`: acknowledged by `revoke_and_ack`.
 * `revoke_and_ack`: acknowledged by `commitment_signed` or `closing_signed`
 * `shutdown`: acknowledged by `closing_signed`.
 
-The last `closing_signed` (if any) must always be retransmitted, as there
-is no explicit acknowledgment.
-
-Before retransmitting `commitment_signed`, the node MUST send
-appropriate `update_` messages (the other node will have forgotten
-them, as required above).
+On reconnection, the node MUST retransmit all `update_` messages, and
```

Comment:
I agree with cdecker here. If specification complaint nodes simply replay the last commitment exchange leading up to the current state, then it seems that they only need to remember the delta between the two states (the `update_*` messages). This seems much simpler to implement correctly compared to allow desynchronization, and attempting to selectively reconcile the divergence. 

With this in mind, from my PoV, the storage requirements are:
   * A delta package of state `#N-1` where the current state number if `N` for the _remote_ peer
   * A delta package is created _locally_ before sending the `commit_sig` message to extend the remote peer's commitment chain. 
 
Upon restart, the local peer retransmits the following:

  * The delta package for state `N-1` for the commitment transaction of the _remote_ peer

I think it's as simple as that? I think specifying retransmission for the state machine distinct from this "acknowledge by" table will make it easer to new comers to quickly digest the details of how retransmission is handled. If the above works, then it seems that the counters themselves actually don't need to be persistent? As we don't care about filtering out duplicates. 

When processing a delta package, you roll back your state to the prior state, apply the package, then if the resulting state matches, send the _prior_ revoke message, and you're done. Otherwise, you missed some messages and should accept this new state.